index.rb

```diff
@@ -1,160 +1,262 @@
- #Search "ruby understanding self and main"
- 
- 
- 
- module Enumerable
-  def my_each
-  
-    for i in self 
-       yield(i)
+module Enumerable
+  def my_each(argument = nil)
+    if block_given?
+      each do |i|
+        if !argument.nil?
+          argument.call(i)
+        else
+          yield(i)
+        end
+      end
+    else
+      self
     end
-
   end
 
-  def my_each_with_index
-  
-    for i in self
-      yield(i, self.find_index(i))
+  def my_each_with_index(argument = nil)
+    i = 0
+    if block_given?
+      length.times do
+        if !argument.nil?
+          argument.call(self[i], i)
+        else
+          yield(self[i], i)
+        end
+        i += 1
+      end
+    else
+      self
     end
```

Comment:
- [ ] my_each_with_index should return an Enumerator if no block is given.
- [ ] my_each_with_index when a block is given when self is an array returns the array itself after calling the given block once for each element in self. that means you have to return the original array after operating on it.
- [ ] my_each_with_index when a block is given when self is a range returns the range itself after calling the given block once for each element in self. that means it should be able to accept ranges.
- [ ] my_each_with_index when a block is given when self is a hash behaves like #each_with_index calls the given block once for each element in self. that means it should accept hashed and operate correctly on them.